pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,369 @@
+			podVolume: v1.Volume{
+				Name: volumeName2,
+				VolumeSource: v1.VolumeSource{
+					PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+						ClaimName: "test-pv",
+						ReadOnly:  false,
+					},
+				},
+			},
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "test-pv",
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+				},
+			},
+			wantPersistentVolume: v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: volumeName2,
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						ScaleIO: &v1.ScaleIOPersistentVolumeSource{},
+					},
+					ClaimRef: &objectReference2,
+				},
+			},
+		},
+		{
+			desc:           "pv type that supports csi migration",
+			seupNodeName:   testNodeName,
+			createNodeName: nodeName,
+			podVolume:      volumeExampleWithCSI,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "claim-name",
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+				},
+			},
+			wantPersistentVolume: pvExampleWithCSI,
+		},
+		{
+			desc:           "CSINode not found for a volume type that supports csi migration",
+			seupNodeName:   testNodeName,
+			createNodeName: kubetypes.NodeName("another-node"),
+			podVolume:      volumeExampleWithCSI,
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kube-apiserver",
+					Namespace: "default",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{
+						{
+							VolumeSource: v1.VolumeSource{
+								HostPath: hostPath,
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: "claim-name",
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErrorMessage: "csiNode \"another-node\" not found",
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			plugMgr, intreeToCSITranslator, csiTranslator, pvLister, pvcLister := setup(test.seupNodeName, t)
+			actualSpec, err := CreateVolumeSpec(test.podVolume, test.pod, test.createNodeName, plugMgr, pvcLister, pvLister, intreeToCSITranslator, csiTranslator)
+
+			if (len(test.wantErrorMessage) > 0 && err == nil) || (err != nil && !strings.Contains(err.Error(), test.wantErrorMessage)) {
+				t.Errorf("got err %v, want err with message %v", err, test.wantErrorMessage)
+			}
+			if !reflect.DeepEqual(emptyPersistentVolume, test.wantPersistentVolume) {
```

Comment:
I think you can just make the `want*` variables pointers, and then you can do comparisions without this check.

Reply:
There is struct within a struct, that is why the comparison == did not work.

Reply:
yes you still need deep equal but I think you could eliminate the top level comparison on an empty struct if you used pointers.